### ISSUE TITLE
Check `shielded_hash` inside `Transfer`s

### ISSUE THREAD

**Issue by murisi** [Zondax/ledger-namada#48]
The `v->transfer.shielded_hash` field is used by the Namada protocol to identify which MASP `Transaction` object to extract from the `Tx`. The hardware wallet currently ignores the `v->transfer.shielded_hash` field both in printing and signing a MASP `Transfer`. (See https://github.com/Zondax/ledger-namada/blob/ca9da5e0129f676934809f58acffbd855d280664/app/src/parser_impl_txn.c#L527 .) This is problematic because a malicious client could potentially tamper with the field and get the hardware wallet to sign over the wrong `shielded_hash`.

If the `v->transfer.shielded_hash` field is present, we should ensure that it value equals `SHA-256(0x04 || txid)` where `txid` is the transaction ID of the MASP `Transaction` inside the `MaspTx` section. If there is no match here, the transaction is invalid and should not be printed nor be signable.

Less importantly, the Ledger app should confirm that the `target_hash` inside the `masp_builder_section_t` points to the MASP section that it provides the plaintexts for. This `target_hash` should also contain the value `SHA-256(0x04 || txid)` where `txid` is the transaction ID of the MASP `Transaction` inside the `MaspTx` section it points to.

**Comment by chcmedeiros** [comment]
Hey @murisi the `txid` needs to be computed has the` blake2b((personalization || consensus_branch_id), header_hash, transparent_hash, sapling_hash)` am I correct ?

**Comment by murisi** [comment]
> Hey @murisi the `txid` needs to be computed has the` blake2b((personalization || consensus_branch_id), header_hash, transparent_hash, sapling_hash)` am I correct ?

Yes, exactly. Is the computed TxId not matching what's contained in the test vectors?

**Comment by chcmedeiros** [comment]
Unfortely yes.  Do you think there is a way you can get me just a single testvector with the expected values of header_hash, transparent_hash, sapling_hash, blake2b((personalization || consensus_branch_id). So I can debug the different hashes I am computing with expected values?

**Comment by murisi** [comment]
> Unfortely yes. Do you think there is a way you can get me just a single testvector with the expected values of header_hash, transparent_hash, sapling_hash, blake2b((personalization || consensus_branch_id). So I can debug the different hashes I am computing with expected values?

The hash data is a little awkward to access, but I'll get your test vector by tomorrow.

**Comment by murisi** [comment]
> Unfortely yes. Do you think there is a way you can get me just a single testvector with the expected values of header_hash, transparent_hash, sapling_hash, blake2b((personalization || consensus_branch_id). So I can debug the different hashes I am computing with expected values?

See attached some test vectors with the hashes.
[tx_digests.txt](https://github.com/user-attachments/files/16100056/tx_digests.txt)

**Comment by chcmedeiros** [comment]
Thanks!

**Comment by murisi** [comment]
Hi @chcmedeiros . Please find attached the debug file and the debug hashes.
[masp_digests.txt](https://github.com/user-attachments/files/16505487/masp_digests.txt)
[vectors_masp.json](https://github.com/user-attachments/files/16505491/vectors_masp.json)
[debugs_masp.txt.gz](https://github.com/user-attachments/files/16505500/debugs_masp.txt.gz)
Thanks!

**Comment by abenso** [comment]
Hi @murisi, I'm having a problem computing the sapling_hash. I was working with the `456_Transfer_0` test of [debugs_masp.txt.gz](https://github.com/user-attachments/files/16505500/debugs_masp.txt.gz). In that test, `shielded_spends`, `shielded_converts` and `value_balance` are empty so I'm only focusing on `shielded_outputs`. 
The `sapling_hash`, in this case is:
`blake2b("ZTxIdSaplingHash" || blake2b("ZTxIdSSpendsHash") || blake2b("ZTxIdConvertHash") || shielded_outputs_hash)` 
where 
`shielded_outputs_hash = blake2b("ZTxIdSOutputHash" ||  blake2b(COMPACT_HASH) || blake2b(MEMOS_HASH) || blake2b(NONCOMPACT_HASH_)) `

From all bytes of `shielded_outputs` I'm computing the different hashes like:
![image](https://github.com/user-attachments/assets/59c2aab5-5865-4aa0-a431-6ee20c3e737d)

In the documentation, it says that we should use the first 52 bytes of enc_ciphertext in the COMPACT_HASH, but in the code, you use 84 bytes. I tested with both, but I don't get the same hash that you shared in [masp_digests.txt](https://github.com/user-attachments/files/16505487/masp_digests.txt): `6fda95102b16ce20cd5266b3b974f5373162d252a778c88a1b01c4d39c078cf6` 

Can you tell at a glance if I am doing something wrong?

**Comment by abenso** [comment]
ok, I found the issue in this case. When value_balance is empty, we need to hash a single byte with a value of 0

**Comment by abenso** [comment]
Hi @murisi, I were able to compute what you refer to as a `Signature Digest` in [masp_digests.txt](https://github.com/user-attachments/files/16505487/masp_digests.txt) file. Then we computed `shielded_hash` like SHA-256(0x04 || `Signature Digest`).
When I want to compare the computed hashes with the hashes coming from the blob, we noticed that: `transaction.sections.maspBuilder.target_hash = transfer.shielded_hash = Signature Digest`, 
so we didn't find any field coming form the blob that is equal to `SHA-256(0x04 || Signature Digest)`

On the other hand, in [masp_digests.txt](https://github.com/user-attachments/files/16505487/masp_digests.txt) file, the `TxId Digest` is the little-endian of `Signature Digest`.

In conclusion, is it necessary to compute SHA-256(0x04 || Signature Digest)? If yes, which part of the blob should we compare this hash to?

**Comment by murisi** [comment]
> Hi @murisi, I were able to compute what you refer to as a `Signature Digest` in [masp_digests.txt](https://github.com/user-attachments/files/16505487/masp_digests.txt) file. Then we computed `shielded_hash` like SHA-256(0x04 || `Signature Digest`). When I want to compare the computed hashes with the hashes coming from the blob, we noticed that: `transaction.sections.maspBuilder.target_hash = transfer.shielded_hash = Signature Digest`, so we didn't find any field coming form the blob that is equal to `SHA-256(0x04 || Signature Digest)`
> 
> On the other hand, in [masp_digests.txt](https://github.com/user-attachments/files/16505487/masp_digests.txt) file, the `TxId Digest` is the little-endian of `Signature Digest`.
> 
> In conclusion, is it necessary to compute SHA-256(0x04 || Signature Digest)? If yes, which part of the blob should we compare this hash to?

Great, thank you for getting the computation working. No, due to some recent changes it is no longer necessary to compute `SHA-256(0x04 || Signature Digest)`.

**Comment by murisi** [comment]
Hi @chcmedeiros . I've been working on integrating MASP hardware wallet signing with the Namada client `namadac`. Sometimes the TxId computations of the hardware wallet and client match, and at other times they do not. I've attached additional test vectors generated by `namadac` where the SUCCESSes refer to the matching case and the FAILUREs refer to the non-matching cases. Can the hardware wallet be made to pass these additional test vectors?
[more_txid_testvectors.txt](https://github.com/user-attachments/files/16963354/more_txid_testvectors.txt)

**Comment by murisi** [comment]
> Hi @chcmedeiros . I've been working on integrating MASP hardware wallet signing with the Namada client `namadac`. Sometimes the TxId computations of the hardware wallet and client match, and at other times they do not. I've attached additional test vectors generated by `namadac` where the SUCCESSes refer to the matching case and the FAILUREs refer to the non-matching cases. Can the hardware wallet be made to pass these additional test vectors? [more_txid_testvectors.txt](https://github.com/user-attachments/files/16963354/more_txid_testvectors.txt)

Hi @chcmedeiros ! I've just rechecked the above test vectors and they seem to be passing now. I suspect that the root cause of the above failures was the return value of `transaction_get_n_spends()` not matching `builder->n_spends`. And this was caused by `namadac` sometimes crashing due to an unrelated protocol error after having generated randomness using the `get_*_randomness` functions and before calling the `sign_masp` function. This meant that in future `namadac` invocations, randomness from old calls was being carried over. Is there a Rust library function that we can call to clear the spend, convert, and output randomness buffers? Ideally we would call this at the beginning of a MASP signing process right before we start calling the `get_*_randomness` functions...

**Comment by chcmedeiros** [comment]
Hey @murisi I was about to write to you, because I was getting the exact values you were sending on the test vectors.
I will call that after we finish the signing process of the transaction. Once you get the wrapper signature I can clean this buffers.

**Comment by murisi** [comment]
> Hey @murisi I was about to write to you, because I was getting the exact values you were sending on the test vectors. I will call that after we finish the signing process of the transaction. Once you get the wrapper signature I can clean this buffers.

Hi @chcmedeiros . Thanks. Would doing this work in those instances where the signing process is somehow interrupted before get to the stage of getting the wrapper signature? In the current `namadac` setup, the signing process can potentially be aborted before `sign_masp` is called or the wrapper signature is requested. `(If` there is an interruption, `namadac` would lose/forget the randomness parameters that have been generated so far...)

**Comment by chcmedeiros** [comment]
Hi @murisi. It will work in case of any step failing on the app and also will reset in the end of sign_masp. Since there is the possibility of you interrupting the process, I will create rust method for cleaning the buffers.

**Comment by murisi** [comment]
> Since there is the possibility of you interrupting the process

Exactly, this is the issue. Thanks!
